processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java

```diff
@@ -680,6 +690,15 @@ public ReadAccessor getReadAccessor(String propertyName, boolean allowedMapToBea
                 .orElse( null );
             return new MapValueAccessor( getMethod, typeParameters.get( 1 ).getTypeMirror(), propertyName );
         }
+        else if ( propertyName.equals( "?" ) && isOptionalType() ) {
```

Comment:
I might consider just throwing a error here if we encounter a `?` but it is not an `Optional` type.